contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,480 @@
+`kubelet` to adjust the Burstable and BestEffort cgroups from consuming memory
+that was requested by a higher QoS class. This increases the risk
+of inducing OOM on BestEffort and Burstable workloads in favor of increasing
+memory resource guarantees for Guaranteed and Burstable workloads.  A value of
+`--experimental-qos-reserved=memory=0%` will allow a Burstable
+and BestEffort QoS sandbox to consume up to the full node allocatable amount if
+available, but increases the risk that a Guaranteed workload will not have
+access to requested memory.
+
+Since memory is an incompressible resource, it is possible that a QoS
+level cgroup sandbox may not be able to reduce memory usage below the
+value specified in the heuristic described earlier during pod admission
+and pod termination.
+
+As a result, the `kubelet` runs a periodic thread to attempt to converge
+to this desired state from the above heuristic.  If unreclaimable memory
+usage has exceeded the desired limit for the sandbox, the `kubelet` will
+attempt to set the effective limit near the current usage to put pressure
+on the QoS cgroup sandbox and prevent further consumption.
+
+The `kubelet` will not wait for the QoS cgroup memory limit to converge
+to the desired state prior to execution of the pod, but it will always
+attempt to cap the existing usage of QoS cgroup sandboxes in lower tiers.
+This does mean that the new pod could induce an OOM event at the `ROOT`
+cgroup, but ideally per our QoS design, the oom_killer targets a pod
+in a lower QoS class, or eviction evicts a lower QoS pod.  The periodic
+task is then able to converge to the steady desired state so any future
+pods in a lower QoS class do not impact the pod at a higher QoS class.
+
+Adjusting the memory limits for the QoS level cgroup sandbox carries
+greater risk given the incompressible nature of memory.  As a result,
+we are not enabling this function by default, but would like operators
+that want to value resource priority over resource utilization to gather
+real-world feedback on its utility.
+
+As a best practice, oeprators that want to provide a similar resource
+reservation model for Guaranteed pods as we offer via enforcement of
+node allocatable are encouraged to schedule their Guaranteed pods first
+as it will ensure the Burstable and BestEffort tiers have had their QoS
+memory limits appropriately ajdusted before taking unbounded workload on
+node.
+
+## Memory backed volumes
+
+The pod level cgroup ensures that any writes to a memory backed volume
+are correctly charged to the pod sandbox even when a container process
+in the pod restarts.
+
+All memory backed volumes are removed when a pod reaches a terminal state.
+
+The `kubelet` verifies that a pod's cgroup is deleted from the
+host before deleting a pod from the API server as part of the graceful
+deletion process.
+
+This ensures resource consumption associated with those volumes are not
```

Comment:
This is not totally true because we do not wait for the pod level cgroup usage to drop to `0` before deleting it.
I feel this is a TODO.